### PR TITLE
gui2: implement widget groups

### PR DESCRIPTION
**Pull request by Vultraz**
A group is a window-unique named vector of widgets. Different widget classes can
use tgroup::execute_group_actions() to iterate over the group members, and pass
a setter function of their own to operate on each group member.

Currently, the code is specialized for the tselectable_ class, but could be
expanded later for other base types if necessary. Out of the tselectable_
inheritors, only ttoggle_button makes use of groups so far (for radio groups).

### CODE REVIEW

**Comment by Vultraz** [comment]
Don't merge this yet. Much work still to do.

**Comment by CelticMinstrel** [comment]
I've already said this in IRC, but... the design here is convoluted and flawed, failing to satisfy required constraints in some cases. The general idea is sound, but it needs a little redesign. Some specification might help, or a minimal example of groups used in an existing dialog. (In other words, not the preferences dialog. Something small.)

**Comment by Vultraz** [comment]
I can't come up with any good answers either, but I hesitate of specializing this to ttoggle_button without moving most of the internals to that class.

**Comment by gfgtdf** [comment]
Hmm i think it makes sense to allow togglepanels too, One could for example use that group to make togglepanels behave like a on a listbox but without forcing the list layout (maybe have a nxm grid of togglepanels, i use a similar thing in my addon).

**Comment by Vultraz** [comment]
Ready for review again.

File: src/gui/dialogs/preferences_dialog.cpp [hunk 1]
```diff
@@ -248,22 +248,24 @@ void tpreferences::setup_combobox(
...
 
+template <typename T, typename G>
 void tpreferences::setup_radio_toggle(
 		const std::string& toggle_id,
-		LOBBY_JOINS enum_value,
-		int start_value,
-		std::vector<std::pair<ttoggle_button*, int> >& vec,
+		const T& enum_value,
+		const int start_value,
+		G& group,
```
Comment (CelticMinstrel) [inline src/gui/dialogs/preferences_dialog.cpp]: I think you actually don't need the second template argument here - instead, you can use `tgroup<T>`.

File: src/gui/widgets/group.hpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+	void group_operator()
+	{
+		FOREACH(AUTO& member, members())
+		{
+			member.first->set_value(false);
+		}
+	}
+
+	T get_active_member_value()
```
Comment (CelticMinstrel) [inline src/gui/widgets/group.hpp]: I suggest adding `set_active_member_value` too.

File: src/gui/dialogs/preferences_dialog.cpp (hunk 1 above)
Comment (Vultraz) [inline src/gui/dialogs/preferences_dialog.cpp]: Done

File: src/gui/widgets/group.hpp (hunk 2 above)
Comment (Vultraz) [inline src/gui/widgets/group.hpp]: Hm... how would that work? Are you referring to setting the second member of the pair, or setting its t/f value?

Comment (CelticMinstrel) [inline src/gui/widgets/group.hpp]: It would take a T as input and set only the widget associated to that value.

Comment (Vultraz) [inline src/gui/widgets/group.hpp]: That has nothing to do with the active member, so I'll call it something else.

**Pull request merged**